### ISSUE TITLE
Training ARIMA Model on the 'Processed_Indoor_Data.csv" file

### ISSUE THREAD

**Issue by Maiskhraishi** [Maiskhraishi/IoT-AQI-Prediction#11]
The ARIMA (AutoRegressive Integrated Moving Average):
This model is a time series forecasting technique used to predict future values based on past data. In our project, it is used to predict the Air Quality Index (AQI) based on historical air quality data.

1. #Import dependencies

import numpy as np
import statsmodels.api as sm
from statsmodels.tsa.arima.model import ARIMA
import matplotlib.pyplot as plt

2. **#Converted Time Column to DateTime Format**

# Ensured 'created_at' is in datetime format

df['created_at'] = pd.to_datetime(df['created_at'])
df.set_index('created_at', inplace=True)

# Sort by date (important for time-series modeling)
df = df.sort_index()

3.Select AQI Column for Forecasting
# Selecting AQI column for prediction 

aqi_column = 'AQI'  
data = df[aqi_column]

# Checking for missing values
data = data.ffill()  # Forwarded-fill missing values
print(data)

4.#Training the ARIMA Model
# Train ARIMA Model (adjust order based on dataset behavior)
model = ARIMA(data, order=(5, 1, 0))  # (p,d,q) values can be tuned
model_fit = model.fit()

# Display summary of model
print(model_fit.summary())

5. #Forecasting Future AQI Values
# Forecast next 10 time points
forecast = model_fit.forecast(steps=10)
print(forecast)

#Predicting AQI Values
# Predicting next 10 time points
predicted = model_fit.forecast(steps=10)
print(predicted)

# Plot the Predicted values
plt.figure(figsize=(10,5))
plt.plot(data, label="Historical AQI")
plt.plot(pd.date_range(data.index[-1], periods=10, freq='H'), predicted, label="Forecast", color='red')
plt.legend()
plt.title("AQI Prediction using ARIMA")
plt.show()

6. Detect Anomalies in AQI Data

from scipy.stats import zscore

# Computed Z-score for anomaly detection
df['z_score'] = zscore(df[aqi_column])

# Defined threshold (e.g., Z-score above 3 or below -3)
threshold = 3
df['anomaly'] = df['z_score'].apply(lambda x: 1 if abs(x) > threshold else 0)

7. #Visualizing Anomalies

plt.figure(figsize=(12,5))
plt.plot(df.index, df[aqi_column], label="AQI", color="blue")
plt.scatter(df[df['anomaly'] == 1].index, df[df['anomaly'] == 1][aqi_column], color="red", label="Anomalies")
plt.legend()
plt.title("Anomaly Detection in AQI Data")
plt.show()

8. # Saving forecasted and anomaly detection
df.to_csv("aqi_results.csv", index=True)

print("Results saved successfully!")

#RE-TRAINING ARIMA MODEL - AFTER ANOMALY REMOVAL

# Predicting AQI Values (Before Anomaly Detection)
predicted = model_fit.forecast(steps=10)
print(predicted)

# Plot the predicted values (Before Anomaly Detection)
plt.figure(figsize=(10,5))
plt.plot(data, label="Historical AQI")
plt.plot(pd.date_range(data.index[-1], periods=10, freq='H'), predicted, label="Forecast", color='red')
plt.legend()
plt.title("AQI Prediction using ARIMA (Before Anomaly Detection)")
plt.show()

# Anomaly Detection (After Anomaly Detection)
from scipy.stats import zscore

# Compute Z-score for anomaly detection
df['z_score'] = zscore(df[aqi_column])

# Define threshold (e.g., Z-score above 3 or below -3)
threshold = 3
df['anomaly'] = df['z_score'].apply(lambda x: 1 if abs(x) > threshold else 0)

# Plotting the anomalies (After Anomaly Detection)
plt.figure(figsize=(10,5))
plt.plot(df.index, df[aqi_column], label="Historical AQI")
plt.scatter(df.index[df['anomaly'] == 1], df[aqi_column][df['anomaly'] == 1], color='red', label='Anomalies', zorder=5)
plt.legend()
plt.title("AQI with Anomalies Detected (After Anomaly Detection)")
plt.show()

# Compare Results
# Before anomaly detection, you have the ARIMA predicted values, while after anomaly detection, anomalies are flagged.
print("Anomalies Detected:", df['anomaly'].sum())

#COMPARING MODEL BEFORE AND AFTER ANOMALY REMOVAL - #aqi values (with mean_absolute_error, mean_squared_error, #r2_score)
#because these are standard error metrics used in regression and #forecasting tasks to evaluate the quality of a predictive model

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from statsmodels.tsa.arima.model import ARIMA
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

# Actual AQI values for the next 10 time points (Assume this is available)
actual_aqi = df['AQI'].iloc[-10:].values  # Replace with actual AQI values

# Predictions before anomaly removal
pred_before = predicted_df['predicted_AQI'].values

# Predictions after anomaly removal (fill NaNs with interpolated values)
pred_after = cleaned_predicted_df['predicted_AQI'].reindex(predicted_df.index).interpolate().values

# Compute error metrics before anomaly removal
mae_before = mean_absolute_error(actual_aqi, pred_before)
mse_before = mean_squared_error(actual_aqi, pred_before)
r2_before = r2_score(actual_aqi, pred_before)

# Compute error metrics after anomaly removal
mae_after = mean_absolute_error(actual_aqi, pred_after)
mse_after = mean_squared_error(actual_aqi, pred_after)
r2_after = r2_score(actual_aqi, pred_after)

# Display results
print(f"Before Anomaly Removal: MAE = {mae_before:.4f}, MSE = {mse_before:.4f}, R2 Score = {r2_before:.4f}")
print(f"After Anomaly Removal: MAE = {mae_after:.4f}, MSE = {mse_after:.4f}, R2 Score = {r2_after:.4f}")

**Comment by Maiskhraishi** [comment]
[preprocessed_indoor_data.csv](https://github.com/user-attachments/files/19595889/preprocessed_indoor_data.csv)
[aqi_results (1).csv](https://github.com/user-attachments/files/19595888/aqi_results.1.csv)

**Comment by Maiskhraishi** [comment]
https://ml.azure.com/fileexplorerAzNB?wsid=/subscriptions/d12ddd59-df27-4054-b1ee-044c3f86854c/resourcegroups/azureml-rg/providers/Microsoft.MachineLearningServices/workspaces/azureml-ws&tid=b30f4b44-46c6-4070-9997-f87b38d4771c&activeFilePath=Users/60090841/aqi_results.csv

#MICROSOFT AZURE-'Preprocessed_indoor_data' - CODE FILE
